servicetalk-loadbalancer/src/test/java/io/servicetalk/loadbalancer/EagerRoundRobinLoadBalancerTest.java

```diff
@@ -0,0 +1,100 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.loadbalancer;
+
+import org.junit.Test;
+
+import static io.servicetalk.concurrent.internal.DeliberateException.DELIBERATE_EXCEPTION;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.is;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+public class EagerRoundRobinLoadBalancerTest extends RoundRobinLoadBalancerTest {
+
+    @Test
+    public void addressIsAddedTwice() {
+        assertThat(lb.activeAddresses(), is(empty()));
+        sendServiceDiscoveryEvents(upEvent("address-1"));
+        assertThat(lb.activeAddresses(), hasSize(1));
+        sendServiceDiscoveryEvents(upEvent("address-1"));
+        assertThat(lb.activeAddresses(), hasSize(2));
+
+        sendServiceDiscoveryEvents(downEvent("address-1"));
+        assertThat(lb.activeAddresses(), hasSize(1));
+        sendServiceDiscoveryEvents(downEvent("address-1"));
+        assertThat(lb.activeAddresses(), hasSize(0));
+    }
+
+    @Test
+    public void handleDiscoveryEvents() {
+        assertAddresses(lb.activeAddresses(), EMPTY_ARRAY);
+
+        sendServiceDiscoveryEvents(upEvent("address-1"));
+        assertAddresses(lb.activeAddresses(), "address-1");
+
+        sendServiceDiscoveryEvents(downEvent("address-1"));
+        assertAddresses(lb.activeAddresses(), EMPTY_ARRAY);
+
+        sendServiceDiscoveryEvents(upEvent("address-2"));
+        assertAddresses(lb.activeAddresses(), "address-2");
+
+        sendServiceDiscoveryEvents(downEvent("address-3"));
+        assertAddresses(lb.activeAddresses(), "address-2");
+
+        sendServiceDiscoveryEvents(upEvent("address-1"));
+        assertAddresses(lb.activeAddresses(), "address-1", "address-2");
+
+        sendServiceDiscoveryEvents(downEvent("address-1"));
+        assertAddresses(lb.activeAddresses(), "address-2");
+
+        sendServiceDiscoveryEvents(downEvent("address-2"));
+        assertAddresses(lb.activeAddresses(), EMPTY_ARRAY);
+
+        sendServiceDiscoveryEvents(downEvent("address-3"));
+        assertAddresses(lb.activeAddresses(), EMPTY_ARRAY);
+
+        // Let's make sure that an SD failure doesn't compromise LB's internal state
+        sendServiceDiscoveryEvents(upEvent("address-1"));
+        assertAddresses(lb.activeAddresses(), "address-1");
+
+        serviceDiscoveryPublisher.onError(DELIBERATE_EXCEPTION);
+        assertAddresses(lb.activeAddresses(), "address-1");
+    }
+
+    @Test
+    public void hostDownGracefulCloseConnection() throws Exception {
+        sendServiceDiscoveryEvents(upEvent("address-1"));
+        TestLoadBalancedConnection conn = lb.selectConnection(any()).toFuture().get();
+        sendServiceDiscoveryEvents(downEvent("address-1"));
```

Comment:
Consider adding tests that verify:
1. If `address-1` has 2+ connections, all will be closed on "unavailable" event.
2. If RRLB has a few addresses, only connections related to the "unavailable" address close, other remain open.